### PR TITLE
ENRTL Model Part 1

### PR DESCRIPTION
**Pull request by andrewlee94**
## Summary/Motivation:
This adds a module for the generic property package framework that implements the eNRTL model. The first part adds the binary interaction parameters required for the local interaction contribution to the activity coefficient.

## Changes proposed in this PR:
- alpha and tau parameters for eNRTL model
- calculation of X, Y, alpha, G and tau in state blocks
- verification tests for the above

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by andrewlee94** [comment]
On a side note, if anyone feels up to checking the equations, I have a copy of the paper you can cross-reference.

**Review by TimBartholomew — DISMISSED** [review]
LGTM. 

Although I don't know the eNRTL model, so I'm just roughly looking at API rather than the implementation of the model. 

I'm assuming you are planning on adding options for the methods determining alpha and tau, which is why you set up a default and created a separate submethods file, correct?

**Comment by adam-a-a** [comment]
> On a side note, if anyone feels up to checking the equations, I have a copy of the paper you can cross-reference.

I'd be interested in checking out the paper.

File: idaes/generic_models/properties/core/eos/tests/test_enrtl.py [hunk 1]
```diff
@@ -0,0 +1,627 @@
...
+                 exp(-model.params.Liq.alpha["H+, Cl-", "H+, OH-"] *
+                     model.params.Liq.tau["H+, Cl-", "H+, OH-"])))
+
+        # Like species interactions
+        assert ("H2O", "H2O") not in model.state[1].Liq_G
+        assert ("C6H12", "C6H12") not in model.state[1].Liq_G
+        assert ("Na+", "Na+") not in model.state[1].Liq_G
+        assert ("Na+", "H+") not in model.state[1].Liq_G
+        assert ("H+", "Na+") not in model.state[1].Liq_G
+        assert ("h+", "H+") not in model.state[1].Liq_G
```
Comment (adam-a-a) [inline idaes/generic_models/properties/core/eos/tests/test_enrtl.py:593]: Should the lowercase "h+" be "H+"?

Comment (andrewlee94) [reply idaes/generic_models/properties/core/eos/tests/test_enrtl.py:593]: Yes it should.

File: idaes/generic_models/properties/core/eos/enrtl.py [hunk 2]
```diff
@@ -26,32 +27,245 @@
...
+            elif ((i in molecular_set) and
+                    (j in molecular_set)):
+                return alpha_rule(b, pobj, i, j, b.temperature)
+            elif (i in b.params.cation_set and j in molecular_set):
+                return sum(
+                    Y[k] *
+                    alpha_rule(
+                        b, pobj, (i+", "+k), j, b.temperature)
+                    for k in b.params.anion_set)
+            elif (j in b.params.cation_set and i in molecular_set):
```
Comment (Andresj89) [inline idaes/generic_models/properties/core/eos/enrtl.py:148]: @andrewlee94 just out of curiosity, is there any particular reason for changing the order of the indices `i` and `j`?

File: idaes/generic_models/properties/core/eos/enrtl.py [hunk 3]
```diff
@@ -26,32 +27,245 @@
...
+                  (i in b.params.anion_set and j in b.params.anion_set)):
+                # No like species interactions
+                return Expression.Skip
+            elif ((i in molecular_set) and
+                    (j in molecular_set)):
+                return tau_rule(b, pobj, i, j, b.temperature)
+            else:
+                alpha = getattr(b, pname+"_alpha")
+                G = getattr(b, pname+"_G")
+                return log(G[i, j])/alpha[i, j]
```
Comment (Andresj89) [inline idaes/generic_models/properties/core/eos/enrtl.py:258]: @andrewlee94 is the "minus" included implicitly here?

Comment (andrewlee94) [reply idaes/generic_models/properties/core/eos/enrtl.py:258]: No, that probably should be fixed.

**Review by Andresj89 — APPROVED** [review]
LGTM

**Review by TimBartholomew — APPROVED** [review]

File: idaes/generic_models/properties/core/eos/enrtl.py [hunk 4]
```diff
@@ -26,32 +27,245 @@
...
+            else:
+                raise BurntToast(
+                    "{} eNRTL model encountered unexpected component pair {}."
+                    .format(b.name, (i, j)))
+        b.add_component(pname+"_G",
+                        Expression(b.params.true_species_set,
+                                   b.params.true_species_set,
+                                   rule=rule_G_expr,
+                                   doc="Local interaction G term"))
+
```
Comment (adam-a-a) [inline idaes/generic_models/properties/core/eos/enrtl.py:241]: I may have missed this, but is the long-range interaction term also accounted for? If it isn't accounted for here, is there some reasoning for excluding the long-range G term?

**Review by adam-a-a — APPROVED** [review]
LGTM

I just reread the summary/motivation and realized that this submission was meant to account for local and not the long-range interaction terms for excess Gibbs energy.

File: idaes/generic_models/properties/core/eos/enrtl.py (hunk 4 above)
Comment (andrewlee94) [reply idaes/generic_models/properties/core/eos/enrtl.py:241]: Yes - the long range contribution is in the next PR. And actually, so is the local contribution - this only sets up the parameters required.

**Pull request merged**